tests/php/Security/PasswordExpirationMiddlewareTest.php

```diff
@@ -0,0 +1,517 @@
+        $request->method('getAcceptMimetypes')->will($this->returnValue(['*/*']));
+        $session = $request->getSession();
+
+        $a->setRedirect($session, '/redirect-address-custom');
+
+        $executed = false;
+        $activeMember = null;
+        Security::setCurrentUser($member);
+        $response = $a->process($request, static function () use (&$executed, &$activeMember) {
+            $executed = true;
+            $activeMember = Security::getCurrentUser();
+        });
+
+        $this->assertTrue($executed);
+        $this->assertEquals($member, $activeMember);
+
+        // Check 2: custom allowed url
+        $request = $this->buildRequestMock('/allowed-address-custom');
+        $request->method('getAcceptMimetypes')->will($this->returnValue(['*/*']));
+        $session = $request->getSession();
+
+        $a->setRedirect($session, '/redirect-address-custom');
+        $a->whitelistUrl($session, '/allowed-address-custom');
+
+        $executed = false;
+        $activeMember = null;
+        $response = $a->process($request, static function () use (&$executed, &$activeMember) {
+            $executed = true;
+            $activeMember = Security::getCurrentUser();
+        });
+
+        $this->assertTrue($executed);
+        $this->assertEquals($member, $activeMember);
+
+        // Check 3: configured merged with custom
+        $request = $this->buildRequestMock('/allowed-address-configured');
+        $request->method('getAcceptMimetypes')->will($this->returnValue(['*/*']));
+        $session = $request->getSession();
+
+        $a->setRedirect($session, '/redirect-address-custom');
+        $a->whitelistUrl($session, '/allowed-address-custom');
+
+        $executed = false;
+        $activeMember = null;
+        $response = $a->process($request, static function () use (&$executed, &$activeMember) {
+            $executed = true;
+            $activeMember = Security::getCurrentUser();
+        });
+
+        $this->assertTrue($executed);
+        $this->assertEquals($member, $activeMember);
+
+        // Check 4: not allowed
+        $request = $this->buildRequestMock('/not-allowed');
+        $request->method('getAcceptMimetypes')->will($this->returnValue(['*/*']));
+        $session = $request->getSession();
+
+        $a->setRedirect($session, '/redirect-address-custom');
+        $a->whitelistUrl($session, '/allowed-address-custom');
+
+        $executed = false;
+        $activeMember = null;
+        $response = $a->process($request, static function () use (&$executed, &$activeMember) {
+            $executed = true;
+            $activeMember = Security::getCurrentUser();
+        });
+
+        $this->assertFalse($executed);
+        $this->assertNull($activeMember);
+
+        $this->assertEquals(302, $response->getStatusCode());
+        $this->assertEquals(Director::absoluteURL('redirect-address-custom'), $response->getHeader('Location'));
```

Comment:
Are you able to split these into four separate test methods? You could use data providers, or just duplicate the middleware creation lines